go/vt/vttablet/tabletmanager/vreplication/vcopier.go

```diff
@@ -253,6 +254,18 @@ func (vc *vcopier) copyTable(ctx context.Context, tableName string, copyState ma
 			buf.Myprintf("update _vt.copy_state set lastpk=%a where vrepl_id=%s and table_name=%s", ":lastpk", strconv.Itoa(int(vc.vr.id)), encodeString(tableName))
 			updateCopyState = buf.ParsedQuery()
 		}
+
+		if rows.Done {
+			log.Infof("Copy of %v finished at lastpk: %v", tableName, bv)
+			buf := sqlparser.NewTrackedBuffer(nil)
+			buf.Myprintf("delete from _vt.copy_state where vrepl_id=%s and table_name=%s", strconv.Itoa(int(vc.vr.id)), encodeString(tableName))
+			if _, err := vc.vr.dbClient.Execute(buf.String()); err != nil {
+				return err
+			}
+
+			return nil
+		}
```

Comment:
i think you want to keep this where it was and instead set a variable outside the scope of this functor to the result of `rows.Done` and return nil there and then check in the lower block for whether that flag was set. This is because by moving this block into the callback from the grpc call, we're changing the semantics of what happens when there's an error in deleting from the copy state table. before, we'd throw an error to the client of `copyTable` then retry, get 0 results from the grpc call _THEN_ delete. Here, we're now erroring out and letting the callback delete for us even if it returned no results.